File: daemons/macos/src/Handlers/SessionJSONLReplay.swift
import Foundation
import Network

enum SessionJSONLReplay {
    static func replay(sessionId: String, to connection: NWConnection) -> Bool {
        if let lines = loadLastTurn(sessionId: sessionId) {
            var seq = 0
            var batch = Data()
            for line in lines {
                if let obj = try? JSONSerialization.jsonObject(with: Data(line.utf8)) {
                    seq += 1
                    let wrapped: [String: Any] = [
                        "event": obj, "seq": seq, "sessionId": sessionId,
                    ]
                    if let payload = try? JSONSerialization.data(withJSONObject: wrapped) {
                        batch.append(payload)
                        batch.append(0x0A)
                    }
                }
            }
            seq += 1
            let tail: [String: Any] = [
                "type": "exit", "code": 0, "seq": seq, "sessionId": sessionId,
            ]
            if let payload = try? JSONSerialization.data(withJSONObject: tail) {
                batch.append(payload)
                batch.append(0x0A)
            }
            connection.send(
                content: batch,
                completion: .contentProcessed { _ in connection.cancel() })
            return true
        }
        return false
    }

    private static func loadLastTurn(sessionId: String) -> [String]? {
        let projects = NSString("~/.claude/projects").expandingTildeInPath
        if let entries = try? FileManager.default.contentsOfDirectory(atPath: projects) {
            for folder in entries {
                let path = "\(projects)/\(folder)/\(sessionId).jsonl"
                if let text = try? String(contentsOfFile: path, encoding: .utf8) {
                    return extractLastTurn(
                        lines: text.split(separator: "\n", omittingEmptySubsequences: true).map(
                            String.init))
                }
            }
        }
        return nil
    }

    private static func extractLastTurn(lines: [String]) -> [String] {
        var cut = 0
        for (index, line) in lines.enumerated().reversed() {
            if isUserPromptEntry(line) {
                cut = index
                break
            }
        }
        return Array(lines[cut..<lines.count])
    }

    private static func isUserPromptEntry(_ line: String) -> Bool {
        if let obj = try? JSONSerialization.jsonObject(with: Data(line.utf8)) as? [String: Any],
            obj["type"] as? String == "user",
            let message = obj["message"] as? [String: Any]
        {
            if let content = message["content"] as? String, !content.isEmpty { return true }
            if let blocks = message["content"] as? [[String: Any]] {
                for block in blocks where block["type"] as? String == "text" { return true }
            }
        }
        return false
    }
}

File: daemons/linux/src/Handlers/SessionJSONLReplay.js
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'

function parsed(line) {
  try {
    return JSON.parse(line)
  } catch {
    return null
  }
}

function isUserPromptEntry(line) {
  const object = parsed(line)
  if (object?.type === 'user' && object.message) {
    if (typeof object.message.content === 'string' && object.message.content.length > 0) {
      return true
    }
    if (Array.isArray(object.message.content)) {
      return object.message.content.some((block) => block.type === 'text')
    }
  }
  return false
}

function extractLastTurn(lines) {
  let cut = 0
  for (let index = lines.length - 1; index >= 0; index -= 1) {
    if (isUserPromptEntry(lines[index])) {
      cut = index
      break
    }
  }
  return lines.slice(cut)
}

function loadLastTurn(sessionId) {
  const projects = path.join(os.homedir(), '.claude', 'projects')
  if (fs.existsSync(projects)) {
    for (const entry of fs.readdirSync(projects, { withFileTypes: true })) {
      if (entry.isDirectory()) {
        const file = path.join(projects, entry.name, `${sessionId}.jsonl`)
        if (fs.existsSync(file)) {
          return extractLastTurn(fs.readFileSync(file, 'utf8').split('\n').filter(Boolean))
        }
      }
    }
  }
  return null
}

export function replay(sessionId, response) {
  const lines = loadLastTurn(sessionId)
  if (lines) {
    let seq = 0
    let batch = ''
    for (const line of lines) {
      const object = parsed(line)
      if (object) {
        seq += 1
        batch += `${JSON.stringify({ event: object, seq, sessionId })}\n`
      }
    }
    batch += `${JSON.stringify({ type: 'exit', code: 0, seq: seq + 1, sessionId })}\n`
    response.end(batch)
    return true
  }
  return false
}
